File: 240626/외판원 순회/traveling-salesman-problem.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;
#define MAX_N 10

int n;
int cost[MAX_N][MAX_N];
bool visited[MAX_N];
vector<int> picked;

int ans = INT_MAX;

void FindMin(int cnt) {
    if(cnt == n) {
        int total_cost = 0;
        for(int i=0; i<(int)picked.size()-1; i++) {
            int curr_cost = cost[picked[i]][picked[i+1]];
            if(curr_cost == 0) return;

            total_cost += curr_cost;
        }

        int additional_cost = cost[picked.back()][0];
        if(additional_cost == 0) return;

        ans = min(ans, total_cost+additional_cost);
        return;
    }

    for(int i=0; i<n; i++) {
        if(visited[i]) continue;
        
        visited[i] = true;
        picked.push_back(i);

        FindMin(cnt+1);

        visited[i] = false;
        picked.pop_back();
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> cost[i][j];
        }
    }

    visited[0] = true;
    picked.push_back(0);
    FindMin(1);

    cout << ans;

    return 0;
}

File: 250404/외판원 순회/traveling-salesman-problem.py
import sys

INT_MAX = sys.maxsize

n = int(input())
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [False]*n
picked = []

ans = INT_MAX

def find_min(idx):
    global ans

    if idx == n:
        total_cost = 0
        for i in range(n-1):
            cur_cost = grid[picked[i]][picked[i+1]]

            if cur_cost == 0:
                return
            
            total_cost += cur_cost
        
        if grid[picked[-1]][0] == 0:
            return
        
        ans = min(ans, total_cost+grid[picked[-1]][0])
        return
    

    for i in range(n):
        if visited[i]:
            continue
        
        visited[i] = True
        picked.append(i)

        find_min(idx+1)

        visited[i] = False
        picked.pop()


visited[0] = True
picked.append(0)
find_min(1)

print(ans)